src/Algebra/Properties/Semigroup/Reasoning.agda

```diff
@@ -0,0 +1,139 @@
+------------------------------------------------------------------------
+-- The Agda standard library
+--
+-- Equational reasoning for semigroups
+-- (Utilities for associativity reasoning, pulling and pushing operations)
+------------------------------------------------------------------------
+
+{-# OPTIONS --cubical-compatible --safe #-}
+
+open import Algebra.Bundles using (Semigroup)
+
+module Algebra.Properties.Semigroup.Reasoning {o ℓ} (S : Semigroup o ℓ) where
+
+open import Algebra.Structures using (IsMagma)
+
+open Semigroup S
+    using (Carrier; _∙_; _≈_; setoid; trans; refl; sym; assoc; ∙-cong; isMagma)
+open import Relation.Binary.Reasoning.Setoid setoid
+open IsMagma isMagma using (∙-congˡ; ∙-congʳ)
+
+private
+ variable
+    u v w x y z : Carrier
+
+module Assoc4  {u v w x : Carrier} where
+  [[u∙v]∙w]∙x≈u∙[[v∙w]∙x] : ((u ∙ v) ∙ w) ∙ x ≈ u ∙ ((v ∙ w) ∙ x)
```

Comment:
While I don't like the 'explicit' names for push/pull, I'm quite fine with these names for the re-associators that work on 4 things!